fl_training.py

```diff
@@ -150,198 +211,146 @@ def compute_test_score(
-
-                # Early stopping
-                print("   Checking if early stopping critera are met:")
-                if is_early_stopping:
-                    # Early stopping parameters
-                    if (
-                        epoch >= constants.PATIENCE
-                        and current_val_loss > global_val_loss[-constants.PATIENCE]
-                    ):
-                        print("      -> Early stopping critera are met, stopping here.")
-                        break
-                    else:
-                        print(
-                            "      -> Early stopping critera are not met, continuing with training."
-                        )
-
-            # Training phase
-            val_acc_list = []
-            acc_list = []
-            for node_index, node in enumerate(node_list):
+    # Train model (iterate for each epoch and mini-batch)
+    print("\n### Training model:")
+    for epoch in range(epoch_count):
+
+        print("\n   Epoch " + str(epoch) + " out of " + str(epoch_count-1) + " total epochs")
+        is_first_epoch = epoch == 0
+        clear_session()
+
+        # Split the train dataset in mini-batches
+        minibatched_x_train, minibatched_y_train = [None] * nodes_count, [None] * nodes_count
+        for node_index, node in enumerate(node_list):
+            minibatched_x_train[node_index], minibatched_y_train[node_index] = split_in_minibatches(minibatch_count, node.x_train, node.y_train)
+
+        # Iterate over mini-batches for training, starting each new iteration with an aggregation of the previous one
+        for minibatch_index in range(minibatch_count):
 
-                print("   Training on node " + str(node))
+            print("\n      Mini-batch " + str(minibatch_index) + " out of " + str(minibatch_count-1) + " total mini-batches")
+            is_first_minibatch = minibatch_index == 0
+
+            # Starting model is the aggregated model from the previous mini-batch iteration
+            if is_first_epoch and is_first_minibatch:
                 node_model = utils.generate_new_cnn_model()
+            else:
+                node_model = build_aggregated_model(aggregate_model_weights(model_list, aggregation_weights))
 
-                # Model weights are the averaged weights
-                if not is_first_epoch:
-                    node_model.set_weights(new_weights)
-                    node_model.compile(
-                        loss=keras.losses.categorical_crossentropy,
-                        optimizer="adam",
-                        metrics=["accuracy"],
-                    )
+            # Iterate over nodes for training each individual model
+            for node_index, node in enumerate(node_list):
 
-                # Train on whole node local data set
+                # Train on node local data set
+                print("         Training on node " + str(node_index) + " - " + str(node))
                 history = node_model.fit(
```

Comment:
`node_model` is not cleared between 2 nodes. It means node_0 trains. Then node_1 trains sequentially on the model trained on node_0. Then node_2 trains sequentially on the model trained on node_0 and node_1 etc.
I think it is valid FL scenario (even probably better than what we did before !) but different to the federated averaging where every node communicates only with a central server/aggregator.

Reply:
In fact we now implemented Sequential FL instead of previous Parallel implementation:
![image](https://user-images.githubusercontent.com/5354301/78222388-76478780-74c5-11ea-8c66-704069ec001b.png)
(taken from Substra whitepaper)